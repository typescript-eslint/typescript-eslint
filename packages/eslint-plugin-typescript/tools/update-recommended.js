"use strict";

const path = require("path");
const fs = require("fs");
const requireIndex = require("requireindex");

/**
 * Generate recommended configuration
 * @returns {void}
 */
function generate() {
    // replace this with Object.entries when node > 8
    const allRules = requireIndex(path.resolve(__dirname, "../lib/rules"));
    const rules = Object.keys(allRules)
        .filter(key => allRules[key].meta.docs.recommended)
        .reduce((config, item) => {
            config[item] = "error";
            return config;
        }, {});

    const filePath = path.resolve(__dirname, "../lib/configs/recommended.json");

    fs.writeFileSync(filePath, `${JSON.stringify({ rules }, null, 4)}\n`);
}

generate();
