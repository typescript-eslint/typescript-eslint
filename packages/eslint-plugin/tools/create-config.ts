/* eslint-disable no-console */

import path from 'path';
import fs from 'fs';
import requireIndex from 'requireindex';

const MAX_RULE_NAME_LENGTH = 33 + 'typescript-eslint/'.length;
const RULE_NAME_PREFIX = '@typescript-eslint';

const all = Object.entries(
  requireIndex(path.resolve(__dirname, '../dist/rules')),
);

const baseConfig = {
  parser: '@typescript-eslint/parser',
  parserOptions: {
    sourceType: 'module',
  },
  plugins: ['@typescript-eslint'],
};

const bannedRules = new Set([
  'camelcase',
  'indent',
  'no-array-constructor',
  'no-unused-vars',
]);

/**
 * Helper function reduces records to key - value pairs.
 * @param config
 * @param entry
 */
const reducer = (
  config: Record<string, string>,
  entry: Record<string, any>,
): Record<string, string> => {
  const key = entry[0];
  const value = entry[1];
  const ruleName = `${RULE_NAME_PREFIX}/${key}`;
  const setting = value.default.meta.docs.recommended;
  const usedSetting = !setting ? 'warn' : setting;

  // having this here is just for output niceness (the keys will be ordered)
  if (bannedRules.has(key)) {
    console.log(key.padEnd(MAX_RULE_NAME_LENGTH), '= off');
    config[key] = 'off';
  }
  console.log(ruleName.padEnd(MAX_RULE_NAME_LENGTH), '=', usedSetting);
  config[ruleName] = usedSetting;

  return config;
};

/**
 * Helper function to check for invalid recommended setting.
 */
function checkValidSettings(): boolean {
  const validSettings = ['error', 'warn', false];
  let result = true;

  all.forEach(entry => {
    const key = entry[0];
    const value = entry[1];
    const setting = value.default.meta.docs.recommended;

    if (!validSettings.includes(setting)) {
      console.error(`ERR! Invalid level for rule ${key}: "${setting}"`);
      result = false;
    }
  });

  return result;
}

/**
 * Helper function generates configuration.
 */
function generate(rules: Record<string, string>, filePath: string): void {
  const config = Object.assign({}, baseConfig, { rules });

  fs.writeFileSync(filePath, `${JSON.stringify(config, null, 2)}\n`);
}

if (checkValidSettings()) {
  console.log('------------------------- all.json -------------------------');
  generate(
    all.reduce(reducer, {}),
    path.resolve(__dirname, '../src/configs/all.json'),
  );

  console.log('--------------------- recommended.json ---------------------');
  const recommendedSettings = ['error', 'warn'];
  generate(
    all
      .filter(entry =>
        recommendedSettings.includes(entry[1].default.meta.docs.recommended),
      )
      .reduce(reducer, {}),
    path.resolve(__dirname, '../src/configs/recommended.json'),
  );
}
