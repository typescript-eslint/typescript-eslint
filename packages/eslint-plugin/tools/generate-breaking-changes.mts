import type { TypeScriptESLintRules } from '@typescript-eslint/eslint-plugin/use-at-your-own-risk/rules';
import { fetch } from 'cross-fetch';
// markdown-table is ESM, hence this file needs to be `.mts`
import { markdownTable } from 'markdown-table';

async function main(): Promise<void> {
  const rulesImport = await import('../src/rules/index.js');
  /*
  weird TS resolution which adds an additional default layer in the type like:
      { default: { default: Rules }}
  instead of just
      { default: Rules }
  @ts-expect-error */
  const rules = rulesImport.default as TypeScriptESLintRules;

  // Annotate which rules are new since the last version
  async function getNewRulesAsOfMajorVersion(
    oldVersion: string,
  ): Promise<Set<string>> {
    // 1. Get the current list of rules (already done)
    const newRuleNames = Object.keys(rules);

    // 2. Retrieve the old version of typescript-eslint from unpkg
    const oldUrl = `https://unpkg.com/@typescript-eslint/eslint-plugin@${oldVersion}/dist/configs/all.js`;
    const oldFileText = await (await fetch(oldUrl)).text();
    const oldObjectText = oldFileText.substring(
      oldFileText.indexOf('{'),
      oldFileText.lastIndexOf('}') + 1,
    );
    // Normally we wouldn't condone using the 'eval' API...
    // But this is an internal-only script and it's the easiest way to convert
    // the JS raw text into a runtime object. 🤷
    let oldRulesObject!: { rules: TypeScriptESLintRules };
    eval('oldRulesObject = ' + oldObjectText);
    const oldRuleNames = new Set(Object.keys(oldRulesObject.rules));

    // 3. Get the keys that exist in (1) (new version) and not (2) (old version)
    return new Set(
      newRuleNames.filter(
        newRuleName => !oldRuleNames.has(`@typescript-eslint/${newRuleName}`),
      ),
    );
  }

  const newRuleNames = await getNewRulesAsOfMajorVersion('5.0.0');

  console.log(`## Table Key

<table>
  <thead>
    <tr>
      <th>Column</th>
      <th>Description</th>
      <th>Emojis</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Status</td>
      <td>Being added, deprecated, or removed</td>
      <td>
        <ul>
          <li>🆕 = newly added to typescript-eslint</li>
          <li>🙅 = to be deprecated in the next major</li>
          <li>💀 = currently deprecated; to be removed in the next version</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>TC</td>
      <td>Requires type checking?</td>
      <td>
        <ul>
          <li>💭 = yes</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Ext</td>
      <td>Extension rule?</td>
      <td>
        <ul>
          <li>🧱 = yes</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Rec'd</td>
      <td>Recommended</td>
      <td>
        <ul>
          <li>➕ = add to recommended this version</li>
          <li>➖️ = remove from recommended this version</li>
          <li>🟩 = stays in recommended this version</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Strict</td>
      <td>Strict</td>
      <td>
        <ul>
        <li>➕ = add to strict this version</li>
        <li>➖️ = remove from strict this version</li>
        <li>🔵 = stays in strict this version</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Style</td>
      <td>Style</td>
      <td>
        <ul>
          <li>➕ = add to stylistic this version</li>
          <li>➖️ = remove from stylistic this version</li>
          <li>🔸 = stays in stylistic this version</li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

## Recommendations Table

> Hint: search for 🆕 to find newly added rules, and ➕ or ➖ to see config changes.
`);

  console.log(
    markdownTable([
      ['Rule', 'Status', 'TC', 'Ext', "Rec'd", 'Strict', 'Style', 'Comment'],
      ...Object.entries(rules).map(([ruleName, { meta }]) => {
        const { deprecated } = meta;
        const { extendsBaseRule, recommended, requiresTypeChecking } =
          meta.docs!;

        return [
          `[\`${ruleName}\`](https://typescript-eslint.io/rules/${ruleName})`,
          newRuleNames.has(ruleName) ? '🆕' : deprecated ? '💀' : '',
          requiresTypeChecking ? '💭' : '',
          extendsBaseRule ? '🧱' : '',
          recommended === 'recommended' ? '🟩' : '',
          recommended === 'strict' ? '🔵' : '',
          recommended === 'stylistic' ? '🔸' : '',
          meta.type === 'layout' ? 'layout 📐' : '(todo)',
        ];
      }),
    ]),
  );
}

main().catch(error => {
  console.error(error);
});
