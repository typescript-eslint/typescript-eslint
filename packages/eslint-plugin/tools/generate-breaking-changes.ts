// eslint-disable-next-line @typescript-eslint/consistent-type-imports
type RulesFile = typeof import('../src/rules');

interface RulesObject {
  default: RulesFile;
}

async function main(): Promise<void> {
  const {
    default: { default: rules },
  } =
    // @ts-expect-error -- We don't support ESM imports of local code yet.
    (await import('../dist/rules/index.js')) as RulesObject;
  const { markdownTable } = await import('markdown-table');
  const { fetch } = await import('cross-fetch');

  const newRuleNames = await getNewRulesAsOfMajorVersion('5.0.0');

  console.log(`## Table Key

<table>
  <thead>
    <tr>
      <th>Column</th>
      <th>Description</th>
      <th>Emojis</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Status</td>
      <td>Being added, deprecated, or removed</td>
      <td>
        <ul>
          <li>🆕 = newly added to typescript-eslint</li>
          <li>🙅 = deprecated in the next major</li>
          <li>➖️ = to be removed from the plugin in the next version</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Ext</td>
      <td>Extension rule?</td>
      <td>
        <ul>
          <li>☑️ = yes</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>R</td>
      <td>Recommended</td>
      <td>
        <ul>
          <li>➕ = add to recommended this version</li>
          <li>⚠️ = recommended as warning</li>
          <li>🛑 = recommended as an error</li>
          <li>➖️ = remove from recommended this version</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>RWT</td>
      <td>Recommended-requiring-type-checking</td>
      <td>
        <ul>
          <li>➕ = add to recommended-with-typechecking this version</li>
          <li>⚠️ = recommended as warning</li>
          <li>🛑 = recommended as an error</li>
          <li>➖️ = remove from recommended this version</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Strict</td>
      <td>Strict</td>
      <td>
        <ul>
          <li>➕ = add to strict this version</li>
          <li>⚠️ = recommended as warning</li>
          <li>➖️ = remove from strict this version</li>
        </ul>
      </td>
  </tr>
  </tbody>
</table>

## Recommendations Table

> Hint: search for 🆕 to find newly added rules, and ➕ or ➖ to see config changes.
`);

  console.log(
    markdownTable([
      ['Rule', 'Status', 'Ext', 'R', 'RWT', 'Strict', 'Comment'],
      ...Object.entries(rules).map(([ruleName, { meta }]) => {
        const { deprecated } = meta;
        const { extendsBaseRule, recommended, requiresTypeChecking } =
          meta.docs!;

        return [
          `[\`${ruleName}\`](https://typescript-eslint.io/rules/${ruleName})`,
          newRuleNames.has(ruleName) ? '🆕' : deprecated ? '🙅' : '',
          extendsBaseRule ? '☑️' : '',
          recommended &&
          ['error', 'warn'].includes(recommended) &&
          !requiresTypeChecking
            ? '🛑'
            : '',
          recommended &&
          ['error', 'warn'].includes(recommended) &&
          requiresTypeChecking
            ? '🛑'
            : '',
          recommended === 'strict' ? '⚠️' : '',
          meta.type === 'layout' ? 'layout 💩' : '(todo)',
        ];
      }),
    ]),
  );

  //  Annotate which rules are new since version 5.0.0
  async function getNewRulesAsOfMajorVersion(
    oldVersion: string,
  ): Promise<Set<string>> {
    // 1. Get the current list of rules (already done)
    const newRuleNames = Object.keys(rules);

    // 2. Use some CDN thing for the 5.X version of typescript-eslint
    const oldUrl = `https://unpkg.com/@typescript-eslint/eslint-plugin@${oldVersion}/dist/configs/all.js`;
    const oldFileText = await (await fetch(oldUrl)).text();
    const oldObjectText = oldFileText.substring(
      oldFileText.indexOf('{'),
      oldFileText.lastIndexOf('}') + 1,
    );
    // Normally we wouldn't condone using the 'eval' API...
    // But this is an internal-only script and it's the easiest way to convert
    // the JS raw text into a runtime object. 🤷
    let oldRulesObject!: { rules: RulesFile };
    eval('oldRulesObject = ' + oldObjectText);
    const oldRuleNames = new Set(Object.keys(oldRulesObject.rules));

    // 3. Get the keys that exist in (1) (new version) and not (2) (old version)
    return new Set(
      newRuleNames.filter(
        newRuleName => !oldRuleNames.has(`@typescript-eslint/${newRuleName}`),
      ),
    );
  }

  await getNewRulesAsOfMajorVersion('5.0.0');
}

main().catch(error => {
  console.error(error);
});
