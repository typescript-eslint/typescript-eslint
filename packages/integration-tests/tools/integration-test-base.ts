import childProcess from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import { promisify } from 'node:util';

import ncp from 'ncp';
import type { DirOptions } from 'tmp';
import tmp from 'tmp';

interface PackageJSON {
  name: string;
  private?: boolean;
  devDependencies: Record<string, string>;
}

const rootPackageJson: PackageJSON = require('../../../package.json');

tmp.setGracefulCleanup();

const copyDir = promisify(ncp.ncp);
const execFile = promisify(childProcess.execFile);
const readFile = promisify(fs.readFile);
const tmpDir = promisify(tmp.dir) as (opts?: DirOptions) => Promise<string>;
const tmpFile = promisify(tmp.file);
const writeFile = promisify(fs.writeFile);

const BASE_DEPENDENCIES: PackageJSON['devDependencies'] = {
  ...global.tseslintPackages,
  eslint: rootPackageJson.devDependencies.eslint,
  typescript: rootPackageJson.devDependencies.typescript,
  jest: rootPackageJson.devDependencies.jest,
};

const FIXTURES_DIR = path.join(__dirname, '..', 'fixtures');
// an env var to persist the temp folder so that it can be inspected for debugging purposes
const KEEP_INTEGRATION_TEST_DIR =
  process.env.KEEP_INTEGRATION_TEST_DIR === 'true';

// make sure that jest doesn't timeout the test
jest.setTimeout(60000);

function integrationTest(
  testName: string,
  testFilename: string,
  executeTest: (testFolder: string) => Promise<void>,
): void {
  const fixture = path.parse(testFilename).name.replace('.test', '');
  describe(fixture, () => {
    const fixtureDir = path.join(FIXTURES_DIR, fixture);

    describe(testName, () => {
      it('should work successfully', async () => {
        const testFolder = await tmpDir({
          keep: KEEP_INTEGRATION_TEST_DIR,
        });
        if (KEEP_INTEGRATION_TEST_DIR) {
          console.error(testFolder);
        }

        // copy the fixture files to the temp folder
        await copyDir(fixtureDir, testFolder);

        // build and write the package.json for the test
        const fixturePackageJson: PackageJSON = await import(
          path.join(fixtureDir, 'package.json')
        );
        await writeFile(
          path.join(testFolder, 'package.json'),
          JSON.stringify({
            private: true,
            ...fixturePackageJson,
            devDependencies: {
              ...BASE_DEPENDENCIES,
              ...fixturePackageJson.devDependencies,
            },
            // ensure everything uses the locally packed versions instead of the NPM versions
            resolutions: {
              ...global.tseslintPackages,
            },
          }),
        );
        // console.log('package.json written.');

        // Ensure yarn uses the node-modules linker and not PnP
        await writeFile(
          path.join(testFolder, '.yarnrc.yml'),
          `nodeLinker: node-modules`,
        );

        await new Promise<void>((resolve, reject) => {
          // we use the non-promise version so we can log everything on error
          childProcess.execFile(
            // we use yarn instead of npm as it will cache the remote packages and
            // make installing things faster
            'yarn',
            // We call explicitly with --no-immutable to prevent errors related to missing lock files in CI
            ['install', '--no-immutable'],
            {
              cwd: testFolder,
            },
            (err, stdout, stderr) => {
              if (err) {
                if (stdout.length > 0) {
                  console.warn(stdout);
                }
                if (stderr.length > 0) {
                  console.error(stderr);
                }
                // childProcess.ExecFileException is an extension of Error
                // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
                reject(err);
              } else {
                resolve();
              }
            },
          );
        });
        // console.log('Install complete.');

        await executeTest(testFolder);
      });

      afterAll(() => {});
    });
  });
}

export function eslintIntegrationTest(
  testFilename: string,
  filesGlob: string,
  flatConfig = false,
): void {
  integrationTest('eslint', testFilename, async testFolder => {
    // lint, outputting to a JSON file
    const outFile = await tmpFile();
    let stderr = '';
    try {
      await execFile(
        'yarn',
        [
          'eslint',
          '--format',
          'json',
          '--output-file',
          outFile,
          '--config',
          flatConfig ? './eslint.config.js' : './.eslintrc.js',
          '--fix-dry-run',
          filesGlob,
        ],
        {
          cwd: testFolder,
        },
      );
    } catch (ex) {
      // we expect eslint will "fail" because we have intentional lint errors

      // useful for debugging
      if (typeof ex === 'object' && ex != null && 'stderr' in ex) {
        stderr = String(ex.stderr);
      }
    }
    // console.log('Lint complete.');
    expect(stderr).toHaveLength(0);

    // assert the linting state is consistent
    const lintOutputRAW = (await readFile(outFile, 'utf8'))
      // clean the output to remove any changing facets so tests are stable
      .replace(
        new RegExp(`"filePath": ?"(/private)?${testFolder}`, 'g'),
        '"filePath": "<root>',
      );
    try {
      const lintOutput = JSON.parse(lintOutputRAW);
      expect(lintOutput).toMatchSnapshot();
    } catch {
      throw new Error(
        `Lint output could not be parsed as JSON: \`${lintOutputRAW}\`.`,
      );
    }
  });
}

export function typescriptIntegrationTest(
  testFilename: string,
  tscArgs: string[],
  assertOutput: (out: string) => void,
): void {
  integrationTest('typescript', testFilename, async testFolder => {
    const [result] = await Promise.allSettled([
      execFile('yarn', ['tsc', '--noEmit', ...tscArgs], {
        cwd: testFolder,
      }),
    ]);

    if (result.status === 'rejected') {
      // this looks weird - but it means that we can show the stdout (the errors)
      // in the test output when typescript fails which helps with debugging
      assertOutput(
        (result.reason as { stdout: string }).stdout.replace(
          // on macos the tmp path might be shown by TS with `/private/`, but
          // the tmp util does not include that prefix folder
          new RegExp(`(/private)?${testFolder}`),
          '/<tmp_folder>',
        ),
      );
    } else {
      // TS logs nothing when it succeeds
      expect(result.value.stdout).toBe('');
      expect(result.value.stderr).toBe('');
    }
  });
}
