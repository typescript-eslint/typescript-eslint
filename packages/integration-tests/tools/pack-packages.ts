/**
 * Pack all of our packages so we can "install" them later.
 * We do this here rather than per test so that we only have
 * to do it once per test run as it takes a decent chunk of
 * time to do.
 * This also ensures all of the tests are guaranteed to run
 * against the exact same version of the package.
 */

import type { TestProject } from 'vitest/node';

import * as child_process from 'node:child_process';
import * as fs from 'node:fs/promises';
import * as os from 'node:os';
import * as path from 'node:path';
import { pathToFileURL } from 'node:url';
import { promisify } from 'node:util';

import rootPackageJson from '../../../package.json';

export const execFile = promisify(child_process.execFile);

interface PackageJSON {
  devDependencies: Record<string, string>;
  name: string;
  private?: boolean;
}

const PACKAGES_DIR = path.resolve(__dirname, '..', '..');

const INTEGRATION_TEST_DIR = path.join(
  os.tmpdir() || os.homedir(),
  'typescript-eslint-integration-tests',
);

const FIXTURES_DIR_BASENAME = 'fixtures';

export const FIXTURES_DESTINATION_DIR = path.join(
  INTEGRATION_TEST_DIR,
  FIXTURES_DIR_BASENAME,
);

const FIXTURES_DIR = path.join(__dirname, '..', FIXTURES_DIR_BASENAME);

const TAR_FOLDER = path.join(INTEGRATION_TEST_DIR, 'tarballs');

export const setup = async (project: TestProject): Promise<void> => {
  const testFileBaseNames = (
    await project.globTestFiles(project.vitest.state.getPaths())
  ).testFiles.map(testFilePath => path.basename(testFilePath, '.test.ts'));

  const PACKAGES = await fs.readdir(PACKAGES_DIR, {
    encoding: 'utf-8',
    withFileTypes: true,
  });

  await fs.mkdir(FIXTURES_DESTINATION_DIR, { recursive: true });

  await fs.mkdir(TAR_FOLDER, { recursive: true });

  const tseslintPackages = Object.fromEntries(
    (
      await Promise.all(
        PACKAGES.map(async ({ name: pkg }) => {
          const packageDir = path.join(PACKAGES_DIR, pkg);
          const packagePath = path.join(packageDir, 'package.json');

          try {
            if (!(await fs.lstat(packagePath)).isFile()) {
              return;
            }
          } catch {
            return;
          }

          const packageJson: PackageJSON = (
            await import(pathToFileURL(packagePath).href, {
              with: { type: 'json' },
            })
          ).default;

          if ('private' in packageJson && packageJson.private === true) {
            return;
          }

          const result = await execFile('npm', ['pack', packageDir], {
            cwd: TAR_FOLDER,
            encoding: 'utf-8',
            shell: true,
          });

          if (typeof result.stdout !== 'string') {
            return;
          }

          const stdoutLines = result.stdout.trim().split('\n');
          const tarball = stdoutLines[stdoutLines.length - 1];

          return [
            packageJson.name,
            `file:${path.join(TAR_FOLDER, tarball)}`,
          ] as const;
        }),
      )
    ).filter(e => e != null),
  );

  const BASE_DEPENDENCIES: PackageJSON['devDependencies'] = {
    ...tseslintPackages,
    eslint: rootPackageJson.devDependencies.eslint,
    typescript: rootPackageJson.devDependencies.typescript,
    vitest: rootPackageJson.devDependencies.vitest,
  };

  const temp = await fs.mkdtemp(path.join(INTEGRATION_TEST_DIR, 'temp'), {
    encoding: 'utf-8',
  });

  await fs.writeFile(
    path.join(temp, 'package.json'),
    JSON.stringify(
      {
        devDependencies: BASE_DEPENDENCIES,
        packageManager: rootPackageJson.packageManager,
        private: true,
        resolutions: tseslintPackages,
      },
      null,
      2,
    ),
    { encoding: 'utf-8' },
  );

  // We install the tarballs here once so that pnpm can cache them globally.
  // This solves 2 problems:
  // 1. Tests can be run concurrently because they won't be trying to install
  //    the same tarballs at the same time.
  // 2. Installing the tarballs for each test becomes much faster as pnpm can
  //    reuse them from its global content-addressable store.
  await execFile('pnpm', ['install', '--no-frozen-lockfile'], {
    cwd: temp,
    shell: true,
  });

  await Promise.all(
    testFileBaseNames.map(async fixture => {
      const testFolder = path.join(FIXTURES_DESTINATION_DIR, fixture);

      const fixtureDir = path.join(FIXTURES_DIR, fixture);

      const fixturePackageJson: PackageJSON = (
        await import(
          pathToFileURL(path.join(fixtureDir, 'package.json')).href,
          { with: { type: 'json' } }
        )
      ).default;

      await fs.cp(fixtureDir, testFolder, { recursive: true });

      await fs.writeFile(
        path.join(testFolder, 'package.json'),
        JSON.stringify(
          {
            private: true,
            ...fixturePackageJson,
            devDependencies: {
              ...BASE_DEPENDENCIES,
              ...fixturePackageJson.devDependencies,
            },

            packageManager: rootPackageJson.packageManager,

            // ensure everything uses the locally packed versions instead of the NPM versions
            resolutions: {
              ...tseslintPackages,
            },
          },
          null,
          2,
        ),
        { encoding: 'utf-8' },
      );

      const { stderr, stdout } = await execFile(
        'pnpm',
        ['install', '--no-frozen-lockfile'],
        {
          cwd: testFolder,
          shell: true,
        },
      );

      if (stderr) {
        console.error(stderr);

        if (stdout) {
          console.log(stdout);
        }
      }
    }),
  );

  await fs.rm(temp, { recursive: true });

  console.log('Finished packing local packages.');
};

export const teardown = async (): Promise<void> => {
  if (process.env.KEEP_INTEGRATION_TEST_DIR !== 'true') {
    await fs.rm(INTEGRATION_TEST_DIR, { recursive: true });
  }
};
