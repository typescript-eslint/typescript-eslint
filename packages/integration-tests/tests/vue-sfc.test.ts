import { eslintIntegrationTest } from '../tools/integration-test-base';

eslintIntegrationTest(__filename, '*.vue');
