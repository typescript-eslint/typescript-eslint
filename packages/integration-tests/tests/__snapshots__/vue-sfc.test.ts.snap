// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vue-sfc eslint should work successfully 1`] = `
[
  {
    "errorCount": 1,
    "fatalErrorCount": 1,
    "filePath": "<root>/Hello.vue",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": [
      {
        "fatal": true,
        "message": "Parsing error: error TS5012: Cannot read file '/usr/linked/tsconfig.json': ENOENT: no such file or directory, open '/usr/linked/tsconfig.json'.",
        "nodeType": null,
        "ruleId": null,
        "severity": 2,
      },
    ],
    "source": "<!-- Hello.vue -->
<template>
  <div>
    <div>
      <!-- !!!!! expected error !!!!! -->
      Hello {{ name as any }}{{ exclamationMarks }}
    </div>
    <button @click="decrement">-</button>
    <button @click="increment">+</button>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
export default Vue.extend({
  props: ['name', 'initialEnthusiasm'],
  data() {
    return {
      enthusiasm: this.initialEnthusiasm,
    };
  },
  methods: {
    increment() {
      this.enthusiasm++;
    },
    decrement() {
      if (this.enthusiasm > 1) {
        this.enthusiasm--;
      }
    },
  },
  computed: {
    exclamationMarks(): any {
      // !!!!! expected error !!!!!
      return Array(this.enthusiasm + 1).join('!');
    },
  },
});
</script>
",
    "suppressedMessages": [],
    "usedDeprecatedRules": [
      {
        "info": {
          "availableUntil": "10.0.0",
          "deprecatedSince": "8.53.0",
          "message": "Formatting rules are being moved out of ESLint core.",
          "replacedBy": [
            {
              "message": "ESLint Stylistic now maintains deprecated stylistic core rules.",
              "plugin": {
                "name": "@stylistic/eslint-plugin-js",
                "url": "https://eslint.style/packages/js",
              },
              "rule": {
                "name": "semi-spacing",
                "url": "https://eslint.style/rules/js/semi-spacing",
              },
              "url": "https://eslint.style/guide/migration",
            },
          ],
          "url": "https://eslint.org/blog/2023/10/deprecating-formatting-rules/",
        },
        "replacedBy": [
          "@stylistic/js/semi-spacing",
        ],
        "ruleId": "semi-spacing",
      },
    ],
    "warningCount": 0,
  },
  {
    "errorCount": 1,
    "fatalErrorCount": 1,
    "filePath": "<root>/Utility.vue",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": [
      {
        "fatal": true,
        "message": "Parsing error: error TS5012: Cannot read file '/usr/linked/tsconfig.json': ENOENT: no such file or directory, open '/usr/linked/tsconfig.json'.",
        "nodeType": null,
        "ruleId": null,
        "severity": 2,
      },
    ],
    "source": "<script lang="ts">
/* https://github.com/typescript-eslint/typescript-eslint/issues/2591 */
/* eslint-disable class-methods-use-this */
export default class Utility {
  get a() {
    const list: Array<string> = [];
    return list;
  }

  get b() {
    const a = this.a as Array<string>;
    return a;
  }
}
</script>
",
    "suppressedMessages": [],
    "usedDeprecatedRules": [
      {
        "info": {
          "availableUntil": "10.0.0",
          "deprecatedSince": "8.53.0",
          "message": "Formatting rules are being moved out of ESLint core.",
          "replacedBy": [
            {
              "message": "ESLint Stylistic now maintains deprecated stylistic core rules.",
              "plugin": {
                "name": "@stylistic/eslint-plugin-js",
                "url": "https://eslint.style/packages/js",
              },
              "rule": {
                "name": "semi-spacing",
                "url": "https://eslint.style/rules/js/semi-spacing",
              },
              "url": "https://eslint.style/guide/migration",
            },
          ],
          "url": "https://eslint.org/blog/2023/10/deprecating-formatting-rules/",
        },
        "replacedBy": [
          "@stylistic/js/semi-spacing",
        ],
        "ruleId": "semi-spacing",
      },
    ],
    "warningCount": 0,
  },
  {
    "errorCount": 1,
    "fatalErrorCount": 1,
    "filePath": "<root>/World.vue",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": [
      {
        "fatal": true,
        "message": "Parsing error: error TS5012: Cannot read file '/usr/linked/tsconfig.json': ENOENT: no such file or directory, open '/usr/linked/tsconfig.json'.",
        "nodeType": null,
        "ruleId": null,
        "severity": 2,
      },
    ],
    "source": "<!-- World.vue -->
<template>
  <div>
    <div :id="name">
      {{ enthusiasm }}
    </div>
    <button @click="decrement">-</button>
    <button @click="increment">+</button>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
export default Vue.extend({
  props: ['name', 'initialEnthusiasm'],
  data() {
    return {
      enthusiasm: this.initialEnthusiasm,
    };
  },
  methods: {
    increment() {
      this.enthusiasm++;
    },
    decrement() {
      if (this.enthusiasm > 1) {
        this.enthusiasm--;
      }
    },
  },
});
</script>
",
    "suppressedMessages": [],
    "usedDeprecatedRules": [
      {
        "info": {
          "availableUntil": "10.0.0",
          "deprecatedSince": "8.53.0",
          "message": "Formatting rules are being moved out of ESLint core.",
          "replacedBy": [
            {
              "message": "ESLint Stylistic now maintains deprecated stylistic core rules.",
              "plugin": {
                "name": "@stylistic/eslint-plugin-js",
                "url": "https://eslint.style/packages/js",
              },
              "rule": {
                "name": "semi-spacing",
                "url": "https://eslint.style/rules/js/semi-spacing",
              },
              "url": "https://eslint.style/guide/migration",
            },
          ],
          "url": "https://eslint.org/blog/2023/10/deprecating-formatting-rules/",
        },
        "replacedBy": [
          "@stylistic/js/semi-spacing",
        ],
        "ruleId": "semi-spacing",
      },
    ],
    "warningCount": 0,
  },
]
`;
